### ISSUE TITLE
Speed Ez - Billing issues

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1668]
In GitLab by @kdjstudios on Apr 1, 2020, 11:55

**Submitted by:**  Joann Browne <joann@speedez.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2020-04-01-57879

**Server:** External

**Client/Site:** Speed Ez

**Account:** NA

**Issue:** 

Is something going on with the billing system it was taking like 20 minutes to upload my files and now it says we’re sorry but something went wrong

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 1, 2020, 11:55

Jo Ann Wrote:

> This is crazy. I had the revert the upload Because it said I had $10,000 worth of late fees I have no idea what is going on here but now it has been processing to upload the fees for 25 minutes
>
> I really am having problems billing. I have uploaded and reverted 4X because none of it is correct. It is taking 25 minutes to upload my file each time and then it gives me an ERROR. Also it says I have over $10,000 in late fees which is a very big error. I need help with this right now please

I wrote:

> Thank you for notifying us. We are currently looking into this. 
> 
> In regards to the upload speed, may you please provide the usage file you are uploading to SA Billing. This way our support team can try to replicate the upload speed/latency. 
> 
> In regards to the late fees, may you provide an account or two that you know are charging the late fees incorrectly so our support team may look into those.

Jo ann wrote:
> The late fees are being charged on all of the paid accounts - one example is account 5195 where they paid the entire bill on the 26th of the month.  Please find attached my upload.
>
> I am wondering if the late charges are because I had to change the billing dates from 3/2/2020-4/1/2020 to the correct dates of 3/1/2020-3/31/2020?

[ResourceUsage_03-01_03-31.csv](/uploads/d6b60b23952bf2a9e82915ed84520fb3/ResourceUsage_03-01_03-31.csv)

I wrote:

> Thank you for the usage file. For the moment, we please ask for your team to stop any billing processing while the support team looks further into this. You may notice the support team upload and apply the usage to the billing cycle to attempt to reproduce the issue with speed and to see how the late fees are being generated. We will keep you updated with their progress and results. 
> 
> In regards to changing the dates on this billing cycle; I do not believe that should have any effect on the late fees. As the late fees are based upon the the previous billing cycle's dates and the account's net terms.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 1, 2020, 12:35

@amishra @rranjan @ranand @tim.traylor @smasih 

I was able to check the 5195 account and here is the last invoice.[invoice_SPDZ-32499.pdf](/uploads/12957c4043ecdaf8f650f602fbc6c189/invoice_SPDZ-32499.pdf) The due dates specified on the invoice is 3/31/20, and the payments were received on 3/26/20. So this account should not generate a late fee. 

Please use the usage file above to test the speed and confirm if late fees are being charged incorrectly.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 1, 2020, 13:33

Hey team I also was looking into the latest late fee option/updates from #1678. As I this might be the culprit to the late fee issues. 

It looks like the team may have enabled it by mistake for all sites both internal and external. I only checked a handful of sites, but they all had it enabled. However, it was supposed to be released with it disabled by default. http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1678#note_56665 may someone please check and confirm this was released as disabled and that users actually went in and enabled them. Or if this was released incorrectly, please change these to disabled for all sites.  

Also as stated in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1678#note_55446 it is not supposed to look at invoices previous to the release date. Just want to mention that too.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 1, 2020, 14:22

@kdjstudios @tim.traylor 

Yes, this is raised by the release of #1678. 

>It looks like the team may have enabled it by mistake for all sites both internal and external. I only checked a handful of sites, but they all had it enabled. However, it was supposed to be released with it disabled by default. http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1678#note_56665 may someone please check and confirm this was released as disabled and that users actually went in and enabled them. Or if this was released incorrectly, please change these to disabled for all sites.
* As per the [comment](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1678#note_55473) Created at is the default option for the late charge.

>Also as stated in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1678#note_55446 it is not supposed to look at invoices previous to the release date. Just want to mention that too.
* We have fixed this.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 1, 2020, 14:31

@ranand @tim.traylor 

Thank you, for the update. However the post you reference is previous to the post I referenced so please change this so created at is only enabled then for the internal sites and it is defaulted to off for the external sites. 

So was my assumption that it was looking at previous invoices the cause to the late fees issue then? or is there more?

Also, how long of time is it taking you to upload the usage file for speed ez? I do not see any updates about this. it should not take anywhere near 20 minutes.

EDIT: I also do not see any response on why we did not receive any email errors about this?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 1, 2020, 14:39

I wrote:

> The support team believes they have found the issue with the late fees and are currently working on publishing the updates to resolve that. We expect to have it fixed within the next 30 minutes. We will follow up once this has been completed and you are ready to proceed with your billing.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 1, 2020, 14:42

@kdjstudios @tim.traylor 

We also didn't get any error emails for the hosted sites, though we did get the error regarding #1772. We will look into it tomorrow morning.

The upload process took only around 2 minutes and the apply process took around 4-5 minutes. We tested it against the given upload file.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 1, 2020, 14:52

@ranand @tim.traylor 

Thank you.

I wrote:

> The support team has confirmed that they have released the updates to production. The team did mention that the Upload step took about 2 minutes and the Apply step took about 4 minutes when they were testing; which they believe these are the normal times it should take to process those steps accordingly.
> 
> You should now be able to resume processing your billing. Please let us know if you have any questions or concerns.

Jo Ann wrote:

> Lol 😂 Huge difference from late charges of over 10,000.00 back to $828.82. Good Job Everyone - Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 1, 2020, 14:55

We have set the created at default to false for the late charge calculations on the hosted server.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 1, 2020, 14:55

@ranand @tim.traylor 

I just want to confirm that the late fee issue was only effecting the external server, or has it been effecting the internal server too? If it has been effecting both, we will need to have a list of all sites that have generated invoices with faulty late fees.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 1, 2020, 14:57

@tim.traylor @kdjstudios 

It affected both the servers. We will give you the list tomorrow.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 2, 2020, 08:06

@ranand @tim.traylor 

Please look into this ASAP. We need to insure the billing goes out. I took a quick look at the site settings and they appear to be setup, but it is showing as disconnected. Please look into why it is now showing as disconnected, when it was just tested and working a few weeks ago. #1626

![image](/uploads/77c5796d12df5dd4ae1ad260640b21e7/image.png)

Jo Ann wrote:

> OK Everyone - I just billed and noticed that my emails did NOT go out through my Server like it was supposed to.  Can you tell me whether or not they went out at all?  I did receive an email that the cycle emailed BUT did you guys GHOST my Administration@speedez.com again because no where do I have an event in Microsoft 365 that it went out through my account at all.  Will wait to hear from you.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 2, 2020, 08:19

@kdjstudios @tim.traylor 

Thank you for the update.

* We have run the script on both of the production servers.
 
#### Hosted

* Only one site(**Laser Answering Service**) has created invoice after the 14/03/2020(Late Charge Release date)
* We have 40 accounts which printed late charge invoice line item. Please find the list of accounts and confirm the late charge is valid or invalid.
* If invalid then you can add a negative staged fee on the account with the invalid amount.  

[Hosted](/uploads/e0b62081b7cd9b70a87d6c9f98813e88/final_keener_late_charge_invoive.rb.ods)

### Same for Answernet Server.

Please find the list of accounts that printed the late fee line item.

[new_answernet.csv](/uploads/de4e49c7c40405a77987f73918090009/new_answernet.csv)

Note: We have got an email tomorrow from the hosted server.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 2, 2020, 08:25

@kdjstudios @tim.traylor 

Thank you for the update.

We have tested to enter our **email address** and got **Success** response on the production server. Please look at the screenshot.

![image](/uploads/68ea4ff39921706583b96cad83520095/image.png)

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 2, 2020, 08:28

@amishra 

Pleased confirm then that all of the billing email invoices Jo Ann sent yesterday went out. As she is stating none of them are showing in her email account.

Also please confirm what is causing it to show as disconnected if it is still connected as you stated above. 

In addition, I still do not see a response on why we are not receiving the error emails on the external server.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 2, 2020, 08:37

@kdjstudios  @tim.traylor 

> Pleased confirm then that all of the billing email invoices Jo Ann sent yesterday went out. As she is stating none of them are showing in her email account.

No email sent out because when Jo-Ann had sent the invoice the status was disconnected. So you need to resend all the invoices.


> Also please confirm what is causing it to show as disconnected if it is still connected as you stated above.

* Please let us know what email address they were trying to input the **Send a Test Email** field.
* We just have entered our email address in the **Send a Test Email** field and got a success response. **NOTHING TO CHANGE IN THE CONFIGURATION.**

> in addition, I still do not see a response on why we are not receiving the error emails on the external server.

EDIT: 
* I would like to inform you that we got the email from hosted server tomorrow.

![image](/uploads/32fc1dfbfd8254f9dff6a026e4ebc501/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 2, 2020, 09:41

We need to have a list of only the invalid late charges, not a list of all late charges. http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1771#note_57544

@tim.traylor 

Late fees:
I presume we will need to ask Gary how we want to handle making the corrections on those accounts once we have the updated list. The team confirmed that the script was not run during the release which is what resulted in the late charges being adding to/for past invoices. 
*  Kyle will check all the sites and inform those that are still in draft to hold off on their billing until wed can confirm the late fees are invalid. "We have concerns with additional late charges, please hold off until we are able to confirm."


Error Emails:
The team has confirmed that they are receiving the error emails. However our team is not. They will be attaching these emails and we can have our techs confirm there are any issues with our email sever.
* We did not receive any emails from Jo Ann's interactions. The notice above is from Kyle's interactions. Here is a screenshot of Jo Ann's Error
![image](/uploads/30030327d06e6106dde47e5f2f88ee2b/image.png) 


Email Integration:
*  The client SMTP connection shows that the account is valid.
*  We need to figure out what happened when the emails were attempted to be sent yesterday.
*  We need to track when we are not able to send emails (To the extent that we can) and send a error accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 2, 2020, 12:03

@kdjstudios @tim.traylor 

We have updated the code, can you please resend the invoices and let us know if they are being sent.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 2, 2020, 13:32

I wrote:

> The support team has implemented an update to the processing and is attempting to send the email invoices again. May you please monitor the sent folder on your admin email account and let us know if the emails start showing. 

Jo Ann wrote:

> ![IMG_0885](/uploads/881adc5d075b831c413b451aafd99296/IMG_0885.jpg)


I wrote:

> We just want to confirm, but judging from the photo, it looks like the emails did go out.

Jo Ann worte:

> I got them.

I wrote:

> We are glad they sent. We will follow up tomorrow with more of an in-depth explain on what was occurring.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 2, 2020, 13:42

I was able to look at the internal server and it would appear that all billing cycles are finalized, so I would presume that there are no draft invoices. I do not have access to the logs to determine if sites/users have sent these invoices out to clients. 

Items still needed:

* List of accounts that generated invoices with faulty late charges. Please include on this list if the invoice was sent out also.
* Improve the send emails functionality. 
   * It should take at max 30 seconds to process. It took about 10 minutes. 
   * Display notification in UI if sending emails was unsuccessful. Also send email error. Currently we were receiving a successful message in the UI. 
* Fix why we are not seeing error emails for the "something went wrong" message.

Please let me know if I am missing anything.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 3, 2020, 05:21

@kdjstudios @tim.traylor 

Thank you for the update.

> * List of accounts that generated invoices with faulty late charges. Please include on this list if the invoice was sent out also.
* List of all Invoices with invalid late fees for the **Internal server**: [internal_server_accounts_with_invalid_late_fees.csv](/uploads/9c131895c8d0e5009a13425040b6a1d4/internal_server_accounts_with_invalid_late_fees.csv)
* List of all Invoices with invalid late fees for the **Hosted server**: 
[hosted_server_accounts_with_invalid_late_fees.csv](/uploads/cfa5a128635b82fd2bbdb410e74ee31f/hosted_server_accounts_with_invalid_late_fees.csv)

> * Improve the send email functionality.
>   * It should take at max 30 seconds to process. It took about 10 minutes.
* We need to do R & D on how we can improve the performance and it will take more time to do. I suggest we need to create a separate ticket for improving the performance in SAB.

>   * Display notification in UI if sending emails was unsuccessful. Also, send the email error. Currently, we were receiving a successful message in the UI.
* We need to fix on our development server first or directly on the production server. Please let us know what we need to do.

> Fix why we are not seeing error emails for the "something went wrong" message.

* We looked into the logs and found that they had multiple times click on revert, upload and apply button but did not find the issue why it was shown **something went wrong.**.
* Also, I tried to test with multiple tab/window to replicate that but did not find anything.

* It might happen that time the server was gone down when the user was proceeding the billing cycle.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 3, 2020, 05:47

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 3, 2020, 06:19

@kdjstudios @tim.traylor @Robleski 

The issue of email not being sent which we fixed last night has affected both internal and external servers. The release on 30 March was the cause of this issue. Because of this, the sites which sent invoice emails through billing cycle send email button in the past 3 days are actually not being sent. We have tracked such the sites and they need to resend the invoices(click the send email button for the billing cycles)

* List of sites for the internal server
  * Site: "Triad", BillingCycle: "Master"
  * Site: "Dyad", BillingCycle: "Master"
  * Site: "Billerica, MA (Inbound)", BillingCycle: "Master"
  * Site: "Santa Rosa, CA", BillingCycle: "Master"
  * Site: "Billings, MT", BillingCycle: "Master"

* For external server, there is only site: "Laser Answering Service", BillingCycle: "Monthly"

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 3, 2020, 07:40

@kdjstudios @Robleski 

As we discussed, I've updated the list with which are actually being sent or not.

List of all Invoices with invalid late fees for the **Internal server**:
[internal_server_accounts_with_invalid_late_fees.csv](/uploads/53c25a68102185141426b1451e653c83/internal_server_accounts_with_invalid_late_fees.csv)

List of all Invoices with invalid late fees for the **Hosted server**:
[hosted_server_accounts_with_invalid_late_fees.csv](/uploads/cc9e9a2952f4655fc16d90ed581d6887/hosted_server_accounts_with_invalid_late_fees.csv)

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 3, 2020, 10:20

@ranand @rranjan @amishra @smasih 

### Topics

1. Fixing the email to work at an acceptable speed will be handled on this ticket. My question is why does it now take so long? Before using the SMTP settings at the admin section would only take 30 seconds max. Are we not doing the same processing, but with different login creds? This makes no sense to me why it is taking so long now and why it didn't work to begin with. Please leave a full and detailed explanation explaining what caused it to not work. Please also leave a full and detailed explanation on what we had to change in order to get it working for the time being; and if that is the result in increased lag. 

2. We now found out that the emails didn't go out for other site's. This is not good at all. This reiterates the 4th post on topic too. So which ticket was the cause of this? http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1771#note_57583 Also, what was missed in testing which allowed for this to be released with this bug?

3. Invalid Late Charges: **How are we determining if an invoice was sent?**
   * Internal: 14 sites with 394 accounts. 202 invoices have been sent. 145 invoices have not been sent. This is totaling 8909.72 in invalid charges or roughly a 86% difference/increase. 
   * External: 3 sites with 76 accounts. 65 invoices have been sent. 11 invoices have not been sent. This is totaling 835.52 in invalid charges or roughly a 84% difference/increase. 

4. Make sure "send email" UI notification results are showing correctly. We currently have been displaying successful, even when they failed. Ensure email error message is also sent when failure to send emails. Please do this on development first and test thoroughly!!!!!

5. Fix issue with error emails for the "something did not work" page not sending out.

### Time:

1. Post how much additional time we have already spent on each of these issues (and any I may have missed) with explanation of what was done during that time.
2. Post an estimate on each of the topics above **before** starting to work on them!! We will need approvals. 

@tim.traylor @Robleski 

### Solutions 

1. I would like your input on how to handle re-sending the emails that did not go out. I believe the simple solution is to go into each billing cycle and just click the "Send Emails" button now. This I know will generate the email list receipt that will be sent to ops at the sites, so I presume before we resend we will need to also email each site notifying them.  
2. I would like your input on how we need to handle adjusting the invalid late fees. I would think we will need to have the Aavaz team add manual invoices to the finalized billing cycles for each site with the correction adjustment for each account. We will need to know which billing code to use; who should we ask to provide that code? NOTE: I say the finalized billing cycle because if we add them to the next billing cycle it will result in all of those site's having incorrect reporting, I would believe.

Also we probably may need Gary's input and approval on these items too before we do anything.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 6, 2020, 06:04

@kdjstudios @Robleski 

> Fixing the email to work at an acceptable speed will be handled on this ticket. My question is why does it now take so long? Before using the SMTP settings at the admin section would only take 30 seconds max. Are we not doing the same processing, but with different login creds? This makes no sense to me why it is taking so long now and why it didn't work to begin with. Please leave a full and detailed explanation explaining what caused it to not work. Please also leave a full and detailed explanation on what we had to change in order to get it working for the time being; and if that is the result in increased lag.

We have looked into it. The process of send email is the same. Apart from getting the client SMTP details for the site, the process is the same. The time this method will take is relative to the number of invoices it is sending. We checked that prior to and after the release #122, it takes ~1 sec/Invoice. It's a sequential process, it iterates over the invoices and calls the mailer to actually send the email.

> We now found out that the emails didn't go out for other site's. This is not good at all. This reiterates the 4th post on topic too. So which ticket was the cause of this? http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1771#note_57583 Also, what was missed in testing which allowed for this to be released with this bug?

At the time of development of #524, we commented out the one line which was actually sending the email, to prevent sending of emails at the time of testing. At the time of its release, we missed to uncomment it. Due to which the emails from send email of the billing cycle page, did not actually get sent.

> Invalid Late Charges: **How are we determining if an invoice was sent?**

For every billing cycle, we are recording their activities. the billing cycle invoice emails which were sent before the release of #524 are being sent and the billing cycles that sent their emails between 29 March to 01 April are not being sent, because of the issue in point 2.
 
> Make sure "send email" UI notification results are showing correctly. We currently have been displaying successful, even when they failed. Ensure email error message is also sent when failure to send emails. Please do this on development first and test thoroughly!!!!!

Estimate for this 2-3 hrs.

> Fix issue with error emails for the "something did not work" page not sending out.

We are unable to recreate the issue that why for particularly that 1 event we didn't get the email. We getting other error emails for the hosted server. 

### Time:

Total 21 Hrs has been spent on this ticket

* Fixed Late fees issue on both the servers: 3 hrs

* R&D and the fix of SMTP issue: 5 hrs 

* Writing Script and to get the list of all faulty emails: 6 hrs

* R&D and get the list of sites whose billing cycle invoices didn't send: 3 hrs

* R&D on why we didn't get the error email: 2 hrs

* R&D on the speed of send email: 2 hrs

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 6, 2020, 09:47

@tim.traylor @Robleski 

The script we ran to get the invalid late fees, checks only the last billing cycle invoices. For the SpeedEz invoices, the late fees are based on the last two billing cycles. That's why we have SpeedEz in the invalid late fees sheet.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 6, 2020, 11:14

### Email Problems

For the 5 sites, if they have manually already applied payments to those invoices. Or if they have sent these to clients via the print and mail or pull bills. We will need to ask this to each site.
 * If the site has not made any payments or sent these out, we still have the option to simply revert and fully rerun the billing cycle. This will result in the correct late fees being generated at this point and will also now allow for emails to be sent. This will make the second late fee point invalid.
 * If the site has done this, then we will need to have them approve for us to resend the emails and also to auto add the staged fees for the late charges below. 

### Late Fees

**For the known sites:**

1. Kyle to separate out the list of invalid accounts by each site. Then to send these to each site.
2. Ask for approval to have the Aavaz team fix
3. Ask for which code to use.
4. Staged fee will be added to each account with the full refund/adjustment amount. This will apply to the next billing cycle.
5. Operations if they want can create and finalize these credits onto a manual invoice.


Email template: **We ask for you assistance in resolving this, and we ask that you please review and confirm as quickly as possible so we can apply the corrections before the next billing cycle.**


**The Fix process:**

1. Once approval from ops at each site.
2. Aavaz will put in the code to prevent any of these accounts to charge late fees on the next billing cycle, as implemented the last time.
3.  Add a staged fee to each account in the full Late charge amount.
**Please run these on dev  for each site as we make these updates before releasing them to Production. As we receive approval on each site**

**Unknown sites:**

We will want to have the Aavaz team rerun the check for invalid late fees to ensure there are no additional sites/accounts. 

### Email send speed

1. We will want to test and compare the spends sending from the office 365 and the Answernet servers to determine if there is a noticeable difference in the speed of sending. This should help point out if we have any issues with internal infrastructure.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Apr 7, 2020, 02:41

@kdjstudios @tim.traylor @Robleski 

> We will want to have the Aavaz team rerun the check for invalid late fees to ensure there are no additional sites/accounts.

While doing more research into the issue, we found that the site's with 0 net-term didn't get affected by the issue. So, all the site's with 0 net-term has valid late fees. 
The count of the affected account's by the issue on the internal server is 201. and the count of the affected account's by the issue on the hosted server is 11

* No new account with invalid late fee found after rerun.

* I've removed the account's with 0 net-term. Please find the updated list of accounts

**Internal Server**
[internal_server_accounts_with_invalid_late_fees_1.csv](/uploads/974c2c4d6d7aa197d45df42162186866/internal_server_accounts_with_invalid_late_fees_1.csv)

**Hosted Server**
[hosted_server_accounts_with_invalid_late_fees_1.csv](/uploads/40e8d8610b4032e95edfa2286d9281a6/hosted_server_accounts_with_invalid_late_fees_1.csv)

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 7, 2020, 10:12

@tim.traylor @Robleski 

So I just want to confirm we need to resend the lists to all the sites now, since the first lists were incorrect.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Apr 7, 2020, 11:20

@kdjstudios @Robleski Hi Kyle, if you didn't already get confirmation, yes.  Thx, Tim

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 7, 2020, 11:38

Thanks @tim.traylor 

I will start separating out the lists for each site and following up with the updated lists. I hope to have this completed before the end of day.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 7, 2020, 13:13

Below is a recap of which sites were resent. If there were the same I did not send any update. If they were removed I also sent an update. 

* Atlanta - Same
* Billerica - Resent ** Still waiting for followup on if invoices were sent or payments made.
* Billings - Same
* Columbus - Removed - Completed
* Dyad - Resent ** Still waiting for followup on if invoices were sent or payments made.
* El Paso - Resent
* Portland - Resent
* Sacramento - Same
* Santa Rosa - Same
* Sarasota - Resent *
* Stockton - Resent
* Toronto - Resent
* Triad - Same
* Winnipeg - Resent

Please let me know if there are any other communications needing to be sent out.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 14, 2020, 01:03

@kdjstudios @tim.traylor 

Thank you for the update.

As per my best understanding, noting to do on our end.

so we can update the labels or we can close this.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 14, 2020, 11:23

@amishra @tim.traylor @Robleski 

I believe we are still needing resolutions for topics 1,4, and 5 in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1771#note_57601.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on May 12, 2020, 01:53

@kdjstudios @Robleski 

> Fixing the email to work at an acceptable speed will be handled on this ticket. My question is why does it now take so long? Before using the SMTP settings at the admin section would only take 30 seconds max. Are we not doing the same processing, but with different login creds? This makes no sense to me why it is taking so long now and why it didn't work to begin with. Please leave a full and detailed explanation explaining what caused it to not work. Please also leave a full and detailed explanation on what we had to change in order to get it working for the time being; and if that is the result in increased lag.

* The process of sending emails is the same, we just added the logic to get the site's SMTP credentials if they are available. 
* The amount of time the send email process will take is directly proportional to the number of invoices it needs to send because it's a sequential process.  
* We did R&D on it, it is taking the same amount of time with Gmail SMTP credentials as it was taking before the release of this feature.
* Average time to send the single email is  ~1 sec. For SpeedEz 387 invoice emails are being sent through the send email process, which takes around ~6.45 mins. The time taken is the same prior to the feature release and is the same as the Gmail SMTP server.
* If we want to enhance this feature then we can make it an asynchronous job/background job. 

> 1. Make sure "send email" UI notification results are showing correctly. We currently have been displaying successful, even when they failed. Ensure email error message is also sent when failure to send emails. Please do this on development first and test thoroughly!!!!!

* We will get back to you on this.

> Fix issue with error emails for the "something did not work" page not sending out.

We didn't find anything regarding why the error email not being sent for that particular error after R&D. We are getting all other error emails for all our dev and production servers.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @amishra on May 18, 2020, 03:41

@kdjstudios @Robleski 

> Make sure "send email" UI notification results are showing correctly. We currently have been displaying successful, even when they failed. Ensure email error message is also sent when failure to send emails. Please do this on development first and test thoroughly!!!!!

* We looked into this and our findings are described below:

* There are two way to send an email.
  * **SMTP email(Default)**
  * **PONY GEM**
* We are using **PONY GEM** to send an email.

Let's take an example:

* When we click on the **send email button** from the billing cycle page.
* there is `100` invoices for sending an email.
* **pony has sent 50 invoices after that in specific time period pony is not working and invoice counter will increase by *70*.and again pony is working and rest of invoices will send.**

* Rest 20 invoices are not delivered to the customer.

* There are two way to handle this errors.

1. When **PONY** will not work we will email default SMTP along with details.
1. we add a flag(true/false) and check each invoice status and show the list of invoices details in the UI.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 18, 2020, 14:58

@Robleski 

I am not sure what they are describing above, do you understand what they are saying and what we need to do to insure we receive an error when emails fail to send?

**Comment by kdjstudios** [comment]
In GitLab by @ranand on May 19, 2020, 24:16

@kdjstudios @Robleski 

In the above comment, Ashish is trying to describe the two different approaches of ensuring that if an email doesn't get through the billing cycle send email process then the site manager or the user gets the mail failure notification or email.

**OPTION 1:**
* Billing Send email is the sequential process and it sends the emails through Pony Gem. In this sequential process, if any email didn't go through or the Pony email service stops working then for every such email we will send a mailer failure email to the site manager or the user using rails default mailer services.
* In this process, the user or the site manager will get the same amount of mailer failure emails as the number of emails didn't get through. For Eg. the 10 emails/invoices which didn't go through, the user will get 10 mailer failure emails

**OPTION 2:**
* In the Billing Send email process we will keep track of which all emails didn't actually go through and after this process gets completed, we will send a consolidated email to the site manager or the user with a list of such invoices and their details.
* So, In this process, the user or the site manager will get a single email with the details of invoices for which the emails didn't go through via default rails email service. This email will be sent only after the send email process gets completed.

So, please let us know which option we should go with?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @ranand on May 19, 2020, 06:10

@kdjstudios @Robleski 

In addition to my above comment

**Option 1** will take 3-4 hrs of estimated effort.

**Option 2** will take 5-6 hrs of estimated effort. 

To complete it, we will need the email content, the sender, and the recipient's list information.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 19, 2020, 10:15

@Robleski 

I think we will want to review this with Gary. However, my input would be the Option 2, one email with the list of all the failures. I can easily see something 'going wrong' and a site getting 100+ emails in Option 1, which would most likely be more problematic then helpful. However, I also can foresee ops missing the one email, which may have 100+ accounts on it, but I think that might just be a training ops topic.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on May 26, 2020, 22:29

Updating labels as per last comment.

**Comment by kdjstudios** [comment]
In GitLab by @gregory.robleski on May 27, 2020, 15:52

Recommended option #2 to Gary; Waiting on his review and approval.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 2, 2020, 09:36

This was discussed today and Gary said option 2 is the way to go.

NOTE: He will want to have the error emails go to the Site email, Plus One, Billing Managers, (Possibly Accounting?)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Jun 3, 2020, 22:47

assigned to @rranjan and unassigned @amishra

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 9, 2020, 06:23

@kdjstudios  @Robleski 

Thanks for the update.

We have completed its development. Please provide the email format for the subject, sender, and the recipient's list information. 

We have generated error and tested it by changing in **Pony** configuration file. Do you also want to test it?

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 9, 2020, 08:30

@rranjan @Robleski 

The below should be the email format. If you can please manually trigger this in your testing, I will be able to confirm as I am part of the "apperrors" email list. NOTE: IDK what the **Pony** config is, but I would say please put in a few invalid email addresses and test with those to confirm it actually works with invalid emails. 


* **From:** suppport@sabilling.com
* **To:** site email
* **CC:** plus1@answernet.com, billingmanagers@answernet.com
* **BCC:** apperrors@sahosted.com
* **Subject:** SA Billing - Errors encountered while sending emails
* **Body:**
> Hi [username],
> 
> The following invoices were unable to be delivered to the below accounts:
> 
> * [Account Name] - [Account Number] - [Account Email]: [Failure Reason]
> * [Account Name] - [Account Number] - [Account Email]: [Failure Reason]
> 
> Please manually resend each account's invoice above. Please reply to this email is you have any questions or concerns.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 10, 2020, 08:53

@amishra @rranjan @ranand @Robleski

Please let us know when this is ready for our testing.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 23, 2020, 01:40

@kdjstudios  @Robleski 

Thanks for the update.

We have triggered the email please check and confirm. Also, we have attached the email snapshot.

**Case 1:-** When the mailer server return error

![image](/uploads/709876357c680cc5fdcc0f059c637a22/image.png)

**Case 2:-** When the recipient is blank.

![image](/uploads/c1f68058fa53750983e36f22033d726c/image.png)



**NOTE:-** We are handling the server-side error.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 23, 2020, 13:17

This looks great, I saw the emails. Marking this for announcement.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Jun 27, 2020, 04:52

@kdjstudios  @Robleski 

We have successfully released on both of the production servers.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 30, 2020, 10:22

closed
